machine_learning/neural_network.cpp

```diff
@@ -0,0 +1,707 @@
+/**
+ * @file neural_network.cpp
+ * @author [Deep Raval](https://github.com/imdeep2905)
+ * 
+ * @brief Implementation of [NeuralNetwork (aka Multilayer Perceptron)] (https://en.wikipedia.org/wiki/Multilayer_perceptron). 
+ * 
+ * @details
+ * A multilayer perceptron (MLP) is a class of feedforward artificial neural network (ANN). The term MLP is used ambiguously, 
+ * sometimes loosely to any feedforward ANN, sometimes strictly to refer to networks composed of multiple layers of perceptrons 
+ * (with threshold activation). Multilayer perceptrons are sometimes colloquially referred to as "vanilla" neural networks, 
+ * especially when they have a single hidden layer.
+ * 
+ * An MLP consists of at least three layers of nodes: an input layer, a hidden layer and an output layer. Except for the 
+ * input nodes, each node is a neuron that uses a nonlinear activation function. MLP utilizes a supervised learning technique 
+ * called backpropagation for training. Its multiple layers and non-linear activation distinguish MLP from a linear 
+ * perceptron. It can distinguish data that is not linearly separable.
+ * 
+ * See [Backpropagation](https://en.wikipedia.org/wiki/Backpropagation) for training algorithm.
+ * 
+ * \note This implementation uses mini-batch gradient descent as optimizer and MSE as loss function. Bias is also not included.
+ */
+
+#include "vector_ops.hpp" // Custom header file for vector operations
+
+#include <iostream>
+#include <valarray>
+#include <vector>
+#include <cmath>
+#include <algorithm>
+#include <chrono>
+#include <string>
+#include <fstream>
+#include <sstream>
+#include <cassert>
+
+/** \namespace machine_learning
+ * \brief Machine learning algorithms
+ */
+namespace machine_learning {
+    /** \namespace neural_network
+     * \brief Neural Network or Multilayer Perceptron
+     */   
+    namespace neural_network {
+        /** \namespace activations
+         * \brief various activation functions used in Neural network
+         */         
+        namespace activations {
+            /**
+             * Sigmoid function
+             * @param X Value 
+             * @return Returns sigmoid(x)
+             */  
+            double sigmoid (const double &x) {
+                return 1.0 / (1.0 + std::exp(-x));
+            }
+
+            /**
+             * Derivative of sigmoid function
+             * @param X Value 
+             * @return Returns derivative of sigmoid(x)
+             */  
+            double dsigmoid (const double &x) {
+                return x * (1 - x);
+            }
+
+            /**
+             * Relu function
+             * @param X Value 
+             * @return Returns relu(x)
+             */  
+            double relu (const double &x) {
+                return std::max(0.0, x);
+            }
+
+            /**
+             * Derivative of relu function
+             * @param X Value 
+             * @return Returns derivative of relu(x)
```

Comment:
Same in other parts of the code.
```suggestion
             * @returns derivative of relu(x)
```